src/crop.sol

```diff
@@ -34,15 +34,41 @@ interface ERC20 {
     function decimals() external returns (uint8);
 }
 
+interface EndLike {
+    function vat() external returns(address);
+    function free(bytes32 ilk) external;
+}
+
 contract UrnProxy {
+    address immutable adapter;
     constructor(address vat) public {
         VatLike(vat).hope(msg.sender);
+        adapter = msg.sender;
+    }
+    function free(EndLike end, bytes32 ilk) external {
+        require(msg.sender == adapter);
+        end.free(ilk);
     }
 }
```

Comment:
So I think we can do this in a more general way without introducing a dependency on `end`. I would just have the UrnProxy look like this:

```
contract UrnProxy {
    address immutable vat;
    address immutable owner;
    constructor(address _vat, address _owner) public {
        VatLike(_vat).hope(msg.sender);
        vat = _vat;
        owner = _owner;
    }
    function hope(address usr) external {
        require(VatLike(vat).live() == 0);
        require(msg.sender == owner);
        VatLike(vat).hope(usr);
    }
}
```

This will allow the user to take full control of the `urn`  during global settlement. Liquidations are off at this point so there is no danger of `tack()` attacks as the system is turned off.

An example of doing `end.free(ilk)` with this method:

```
// Settlement occurs vat.live() == 0
(uint256 ink, uint256 art) = vat.urns(ilk, address(myProxy));
myProxy.hope(address(this));
vat.fork(ilk, address(myProxy), address(this), ink, art);
adapter.tack(address(myProxy), address(this), ink);
end.free(ilk);	// Get back your gems here
adapter.exit(address(this), ink);
```

It's a bit of extra work to clear out this collateral, but all things should be do-able with this method as the `urn` is degrading into a regular collateral with this. This would also cover any additional settlement edge case we are missing right now.

Reply:
Yeah, I like that better--it's completely agnostic to the End interface.